### ISSUE TITLE
Repeating INFO logs on calico-policy-controller after resource deletion

### ISSUE THREAD

**Issue by ktsakalozos** [projectcalico/canal#100]
## Expected Behavior
After a resource is deleted the controller should stop trying to delete it.

## Current Behavior
After deleting a namespace or network policy the calico-policy-controller keeps repeating every 10 seconds its last action:
```
2017-09-06 08:32:40,722 7 INFO Handled MODIFIED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:32:42,549 7 INFO Handled MODIFIED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:14,340 7 INFO Handled MODIFIED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:14,643 7 INFO Handled MODIFIED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:22,385 7 INFO Handled MODIFIED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:22,392 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:23,880 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:33:32,410 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:33,909 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:33:33,909 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:33:42,432 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:43,924 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:33:43,924 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:33:52,464 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:33:53,950 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:33:53,950 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:02,493 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:03,975 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:03,975 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:12,523 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:13,996 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:13,996 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:22,555 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:24,015 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:24,015 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:32,570 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:34,038 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:34,038 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:42,604 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:44,062 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:44,062 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:34:52,633 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:34:54,343 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:34:54,343 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:35:02,666 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:35:04,121 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:35:04,121 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:35:12,703 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:35:14,147 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:35:14,147 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
2017-09-06 08:35:22,729 7 INFO Handled DELETED for Pod: ('policy-demo', 'access-2475823348-815fq')
2017-09-06 08:35:24,165 7 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-09-06 08:35:24,165 7 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
...... <forever> ....
```

## Steps to Reproduce (for bugs)
1. Deploy Canal. We used Juju so the deployment was `juju deploy cs:~containers/canonical-kubernetes-canal-4`  
2. Create a namespace and a network policy:
```
> kubectl create ns policy-demo
> kubectl run --namespace=policy-demo nginx --replicas=2 --image=nginx
> kubectl expose --namespace=policy-demo deployment nginx --port=80
> kubectl create -f - <<EOF
kind: NetworkPolicy
apiVersion: extensions/v1beta1
metadata:
  name: default-deny
  namespace: policy-demo
spec:
  podSelector:
EOF
```
3. Delete the network policy
```
> kubectl delete netpol -n policy-demo default-deny
```
4. See the logs of policy controller:
```
> kubectl logs po/calico-policy-controller-1263594417-dql2d -n kube-system -f
```

## Your Environment

* Calico version: 2.5.1
* Flannel version: 0.7.0
* Orchestrator version: Juju 2.2
* Operating System and version: Ubuntu 16.04
* Link to your project (optional):

**Comment by caseydavenport** [comment]
@ktsakalozos hmm, that is odd. I haven't seen that before. 

AFAIK that should only happen when new updates are received from the k8s API. 

Could you try reproducing with DEBUG logging enabled? That might give more clues as to what's going on.

**Comment by heschlie** [comment]
Was able to replicate this in Calico v2.5.1 with calico/kube-policy-controller v0.7.0, but it appears to not be present in the new calico/kube-controllers v1.0.0 in Calico v2.6.2.

Still doing a bit of digging to see why this is happening in v2.5.1.

**Comment by heschlie** [comment]
With debug turned on, this is the section that repeats endlessly:

```
2017-10-19 21:05:09,278 5 DEBUG Emptied the event queue
2017-10-19 21:05:09,278 5 DEBUG Reading from event queue
2017-10-19 21:05:13,380 5 DEBUG Normal read time out for Pod
2017-10-19 21:05:13,380 5 DEBUG Restarting watch on resource: Pod
2017-10-19 21:05:13,380 5 DEBUG Starting watch on: https://kubernetes.default:443/api/v1/watch/pods
2017-10-19 21:05:13,380 5 DEBUG Getting API resources 'https://kubernetes.default:443/api/v1/watch/pods' at version '420'. stream=True
2017-10-19 21:05:13,399 5 DEBUG Watch response for https://kubernetes.default:443/api/v1/watch/pods: <Response [200]>
2017-10-19 21:05:14,396 5 DEBUG Normal read time out for Namespace
2017-10-19 21:05:14,398 5 DEBUG Restarting watch on resource: Namespace
2017-10-19 21:05:14,398 5 DEBUG Starting watch on: https://kubernetes.default:443/api/v1/watch/namespaces
2017-10-19 21:05:14,399 5 DEBUG Getting API resources 'https://kubernetes.default:443/api/v1/watch/namespaces' at version '369'. stream=True
2017-10-19 21:05:14,435 5 DEBUG Watch response for https://kubernetes.default:443/api/v1/watch/namespaces: <Response [200]>
2017-10-19 21:05:19,284 5 DEBUG Normal read time out for NetworkPolicy
2017-10-19 21:05:19,286 5 DEBUG Restarting watch on resource: NetworkPolicy
2017-10-19 21:05:19,287 5 DEBUG Starting watch on: https://kubernetes.default:443/apis/extensions/v1beta1/watch/networkpolicies
2017-10-19 21:05:19,288 5 DEBUG Getting API resources 'https://kubernetes.default:443/apis/extensions/v1beta1/watch/networkpolicies' at version '398'. stream=True
2017-10-19 21:05:19,316 5 DEBUG Watch response for https://kubernetes.default:443/apis/extensions/v1beta1/watch/networkpolicies: <Response [200]>
2017-10-19 21:05:19,317 5 DEBUG Read from API: {"type":"DELETED","object":{"kind":"NetworkPolicy","apiVersion":"extensions/v1beta1","metadata":{"name":"default-deny","namespace":"policy-demo","selfLink":"/apis/extensions/v1beta1/namespaces/policy-demo/networkpolicies/default-deny","uid":"1c5bc843-b511-11e7-912d-0800275f1e48","resourceVersion":"398","generation":1,"creationTimestamp":"2017-10-19T21:04:35Z"},"spec":{"podSelector":{}}}}
2017-10-19 21:05:19,317 5 DEBUG DELETED NetworkPolicy: default-deny to queue (0) (1508447119.32)
2017-10-19 21:05:19,317 5 DEBUG Read event: DELETED, NetworkPolicy, {
  "kind": "NetworkPolicy",
  "spec": {
    "podSelector": {}
  },
  "apiVersion": "extensions/v1beta1",
  "metadata": {
    "name": "default-deny",
    "generation": 1,
    "namespace": "policy-demo",
    "resourceVersion": "398",
    "creationTimestamp": "2017-10-19T21:04:35Z",
    "selfLink": "/apis/extensions/v1beta1/namespaces/policy-demo/networkpolicies/default-deny",
    "uid": "1c5bc843-b511-11e7-912d-0800275f1e48"
  }
}
2017-10-19 21:05:19,317 5 DEBUG Processing 'DELETED' for kind 'NetworkPolicy'
2017-10-19 21:05:19,317 5 DEBUG Looking up handler for event: ('NetworkPolicy', 'DELETED')
2017-10-19 21:05:19,317 5 DEBUG Deleting network policy: policy-demo.default-deny
2017-10-19 21:05:19,318 5 DEBUG Update resourceVersion, was: 398, now: 398
2017-10-19 21:05:19,320 5 INFO Unable to find policy 'policy-demo.default-deny' - already deleted
2017-10-19 21:05:19,320 5 INFO Handled DELETED for NetworkPolicy: ('policy-demo', 'default-deny')
```
We appear to always restart from resource version 398, not sure if we should be moving past it somewhere and we aren't, would need to dig in further. Not sure if this issue exists in any of the older policy controllers yet.

**Comment by heschlie** [comment]
Also seeing similar repeat logs for pods and namespaces in the policy controller.

**Comment by heschlie** [comment]
@ktsakalozos Would you be able to update your Calico related versions in the Canal manifest to the [Calico v2.6.2](https://docs.projectcalico.org/v2.6/releases/) versions? It definitely solves this problem, we are going to patch in a fix to v2.5.x but I do not have an ETA on it.

**Comment by ktsakalozos** [comment]
Thank you

**Comment by caseydavenport** [comment]
This seems to be fixed in Calico v2.6. The kube-policy-controller container is deprecated now, so please update to the newer version. Thanks!
